# Assignment 07 - Python Basics & Data Exploration/Cleaning with Pandas

Dataset: `dataset/superstore_utf8.csv` (Sample Superstore sales data)

Steps covered:
1. Load CSV into a DataFrame
2. Explore data (head/tail, shape, columns, dtypes)
3. Handle missing values (identify, fill/drop)
4. Basic operations (filter rows, select columns)
5. Remove duplicates
6. Create derived column `total_amount = price * quantity`
7. Save cleaned dataset as a new CSV

In [1]:
import pandas as pd

df = pd.read_csv("dataset/superstore_utf8.csv")
print("Loaded dataset with shape:", df.shape)

Loaded dataset with shape: (9994, 21)


## 2. Explore the data

In [2]:
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


In [3]:
df.tail()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
9989,9990,CA-2014-110422,1/21/2014,1/23/2014,Second Class,TB-21400,Tom Boeckenhauer,Consumer,United States,Miami,...,33180,South,FUR-FU-10001889,Furniture,Furnishings,Ultra Door Pull Handle,25.248,3,0.2,4.1028
9990,9991,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,FUR-FU-10000747,Furniture,Furnishings,Tenex B1-RE Series Chair Mats for Low Pile Car...,91.960,2,0.0,15.6332
9991,9992,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,TEC-PH-10003645,Technology,Phones,Aastra 57i VoIP phone,258.576,2,0.2,19.3932
9992,9993,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,OFF-PA-10004041,Office Supplies,Paper,"It's Hot Message Books with Stickers, 2 3/4"" x 5""",29.600,4,0.0,13.3200
9993,9994,CA-2017-119914,5/4/2017,5/9/2017,Second Class,CC-12220,Chris Cortes,Consumer,United States,Westminster,...,92683,West,OFF-AP-10002684,Office Supplies,Appliances,"Acco 7-Outlet Masterpiece Power Center, Wihtou...",243.160,2,0.0,72.9480


In [4]:
print("Shape:", df.shape)
print("\nColumns:", list(df.columns))

Shape: (9994, 21)

Columns: ['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode', 'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State', 'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category', 'Product Name', 'Sales', 'Quantity', 'Discount', 'Profit']


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   object 
 2   Order Date     9994 non-null   object 
 3   Ship Date      9994 non-null   object 
 4   Ship Mode      9994 non-null   object 
 5   Customer ID    9994 non-null   object 
 6   Customer Name  9994 non-null   object 
 7   Segment        9994 non-null   object 
 8   Country        9994 non-null   object 
 9   City           9994 non-null   object 
 10  State          9994 non-null   object 
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   object 
 13  Product ID     9994 non-null   object 
 14  Category       9994 non-null   object 
 15  Sub-Category   9994 non-null   object 
 16  Product Name   9994 non-null   object 
 17  Sales          9994 non-null   float64
 18  Quantity

In [6]:
df.describe()

,Row ID,Postal Code,Sales,Quantity,Discount,Profit
count,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000
mean,4997.500000,55190.379428,229.858001,3.789574,0.156203,28.656896
std,2885.163629,32063.693350,623.245101,2.225110,0.206452,234.260108
min,1.000000,1040.000000,0.444000,1.000000,0.000000,-6599.978000
25%,2499.250000,23223.000000,17.280000,2.000000,0.000000,1.728750
50%,4997.500000,56430.500000,54.490000,3.000000,0.200000,8.666500
75%,7495.750000,90008.000000,209.940000,5.000000,0.200000,29.364000
max,9994.000000,99301.000000,22638.480000,14.000000,0.800000,8399.976000


## 3. Handle missing values

In [7]:
missing_counts = df.isnull().sum()
print("Missing values per column:")
print(missing_counts[missing_counts > 0])

Missing values per column:
Series([], dtype: int64)


In [8]:
# Postal Code is numeric and has a handful of missing values -> fill with 0 (unknown)
# Any row missing a critical identifier (Order ID / Customer ID / Product ID) is dropped
df["Postal Code"] = df["Postal Code"].fillna(0)

critical_cols = ["Order ID", "Customer ID", "Product ID"]
before = len(df)
df = df.dropna(subset=critical_cols)
print(f"Dropped {before - len(df)} rows missing critical identifiers")
print("\nRemaining missing values:\n", df.isnull().sum().sum())

Dropped 0 rows missing critical identifiers

Remaining missing values:
 0


## 4. Basic operations: filter rows, select columns

In [9]:
# Filter: orders from the 'West' region with Sales > 500
west_high_sales = df[(df["Region"] == "West") & (df["Sales"] > 500)]
print("West region orders with Sales > 500:", west_high_sales.shape[0])
west_high_sales.head()

West region orders with Sales > 500: 374


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
7,8,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,TEC-PH-10002275,Technology,Phones,Mitel 5320 IP Phone VoIP phone,907.152,6,0.2,90.7152
10,11,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,FUR-TA-10001539,Furniture,Tables,Chromcraft Rectangular Conference Tables,1706.184,9,0.2,85.3092
11,12,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,TEC-PH-10002033,Technology,Phones,Konftel 250 Conference phone - Charcoal black,911.424,4,0.2,68.3568
24,25,CA-2015-106320,9/25/2015,9/30/2015,Standard Class,EB-13870,Emily Burns,Consumer,United States,Orem,...,84057,West,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,1044.630,3,0.0,240.2649
67,68,CA-2014-106376,12/5/2014,12/10/2014,Standard Class,BS-11590,Brendan Sweed,Corporate,United States,Gilbert,...,85234,West,OFF-AR-10002671,Office Supplies,Art,Hunt BOSTON Model 1606 High-Volume Electric Pe...,1113.024,8,0.2,111.3024


In [10]:
# Select a subset of columns
selected_cols = df[["Order ID", "Customer Name", "Category", "Sales", "Quantity", "Profit"]]
selected_cols.head()

,Order ID,Customer Name,Category,Sales,Quantity,Profit
0,CA-2016-152156,Claire Gute,Furniture,261.9600,2,41.9136
1,CA-2016-152156,Claire Gute,Furniture,731.9400,3,219.5820
2,CA-2016-138688,Darrin Van Huff,Office Supplies,14.6200,2,6.8714
3,US-2015-108966,Sean O'Donnell,Furniture,957.5775,5,-383.0310
4,US-2015-108966,Sean O'Donnell,Office Supplies,22.3680,2,2.5164


## 5. Remove duplicates

In [11]:
dup_count = df.duplicated().sum()
print("Duplicate rows found:", dup_count)
df = df.drop_duplicates()
print("Shape after removing duplicates:", df.shape)

Duplicate rows found: 0
Shape after removing duplicates: (9994, 21)


## 6. Create a derived column: `total_amount = price * quantity`

The dataset has no standalone unit-price column. `Sales` is the line-item total, so unit `price` is derived as `Sales / Quantity`, then `total_amount` is recomputed as `price * Quantity` to demonstrate the derived-column technique.

In [12]:
df["price"] = df["Sales"] / df["Quantity"]
df["total_amount"] = df["price"] * df["Quantity"]
df[["Sales", "Quantity", "price", "total_amount"]].head()

,Sales,Quantity,price,total_amount
0,261.9600,2,130.9800,261.9600
1,731.9400,3,243.9800,731.9400
2,14.6200,2,7.3100,14.6200
3,957.5775,5,191.5155,957.5775
4,22.3680,2,11.1840,22.3680


## 7. Save the cleaned dataset

In [13]:
df.to_csv("cleaned_superstore.csv", index=False)
print("Saved cleaned dataset to cleaned_superstore.csv with shape:", df.shape)

Saved cleaned dataset to cleaned_superstore.csv with shape: (9994, 23)


## Summary

Working through this Superstore dataset turned out to be a good refresher on the basics of Pandas. After loading the CSV, I found a clean dataset right out of the gate — 9,994 orders across 21 columns, with sensible types (dates and IDs as text, sales/profit as floats, quantity as an integer) and not a single missing value anywhere, including `Postal Code`, which I'd expected might need filling. Since there genuinely were no gaps, the fill/drop logic I wrote is there mainly to show the pattern (fill `Postal Code` with 0, drop rows missing an Order/Customer/Product ID) rather than to fix anything real.

Filtering and column selection were straightforward: pulling out West-region orders over $500 in sales surfaced 374 rows, and narrowing the table down to just Order ID, Customer Name, Category, Sales, Quantity, and Profit made it much easier to read. Checking for duplicates came up empty too — `duplicated()` reported zero, so `drop_duplicates()` didn't actually remove anything, but it's a step worth keeping in any cleaning pipeline regardless of whether this particular file needed it.

The one genuine gap in the raw data was a unit price — the file only gives `Sales` (the line total) and `Quantity`. So I derived `price` as `Sales / Quantity` and then built `total_amount = price * Quantity` on top of it, which is exactly what the assignment asked for, even though here it lands back on the same number as `Sales` by construction.

Final result: the same 9,994 rows, now with two new columns (`price` and `total_amount`), saved out to `cleaned_superstore.csv`. Nothing dramatic needed fixing in this dataset, but every step of the standard load → explore → clean → transform → save workflow is there and working.